File: Apps/Tab/Sources/Tab/Models/Entities.swift
import Foundation
import SwiftData

// MARK: - Profile

@Model
final class ProfileEntity {
    #Unique<ProfileEntity>([\.id])

    var id: UUID
    var displayName: String
    var avatarURL: String?
    /// Read cursor for the Activity feed. Unread = activity newer than this.
    var activityLastSeenAt: Date?
    var updatedAt: Date
    var writeID: UUID
    var pushedWriteID: UUID?

    init(
        id: UUID,
        displayName: String,
        avatarURL: String? = nil,
        activityLastSeenAt: Date? = nil,
        updatedAt: Date = .now,
        writeID: UUID = UUID(),
        pushedWriteID: UUID? = nil
    ) {
        self.id = id
        self.displayName = displayName
        self.avatarURL = avatarURL
        self.activityLastSeenAt = activityLastSeenAt
        self.updatedAt = updatedAt
        self.writeID = writeID
        self.pushedWriteID = pushedWriteID
    }
}

// MARK: - Trip

@Model
final class TripEntity {
    #Unique<TripEntity>([\.id])

    var id: UUID
    var name: String
    /// "trip" (a real, user-facing trip) or "non_group" (a hidden shadow group
    /// backing non-group expenses; never shown in the Trips list, server-managed).
    var kind: String = "trip"
    /// Canonical participant-set signature for non-group containers; nil for trips.
    var memberSignature: String?
    var createdByID: UUID
    var lastActivityAt: Date
    var createdAt: Date
    var updatedAt: Date
    var deletedAt: Date?
    var writeID: UUID
    var pushedWriteID: UUID?

    @Relationship(deleteRule: .cascade, inverse: \TripPersonEntity.trip)
    var people: [TripPersonEntity] = []

    @Relationship(deleteRule: .cascade, inverse: \ExpenseEntity.trip)
    var expenses: [ExpenseEntity] = []

    @Relationship(deleteRule: .cascade, inverse: \SettlementEntity.trip)
    var settlements: [SettlementEntity] = []

    var isNonGroup: Bool { kind == "non_group" }

    /// Current members: excludes soft-removed people. Use for member lists,
    /// avatar groups, and new-expense pickers. Removed people stay in `people`
    /// so their names keep resolving in history and balances.
    var activePeople: [TripPersonEntity] { people.filter { $0.removedAt == nil } }

    init(
        id: UUID = UUID(),
        name: String,
        kind: String = "trip",
        memberSignature: String? = nil,
        createdByID: UUID,
        lastActivityAt: Date = .now,
        createdAt: Date = .now,
        updatedAt: Date = .now,
        deletedAt: Date? = nil,
        writeID: UUID = UUID(),
        pushedWriteID: UUID? = nil
    ) {
        self.id = id
        self.name = name
        self.kind = kind
        self.memberSignature = memberSignature
        self.createdByID = createdByID
        self.lastActivityAt = lastActivityAt
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.deletedAt = deletedAt
        self.writeID = writeID
        self.pushedWriteID = pushedWriteID
    }
}

// MARK: - TripPerson

@Model
final class TripPersonEntity {
    #Unique<TripPersonEntity>([\.id])

    var id: UUID
    var userID: UUID?
    var email: String
    var displayName: String
    var invitedByID: UUID?
    var joinedAt: Date?
    /// Membership state, not deletion: a removed person keeps their ledger
    /// rows and balances but drops out of member lists and pickers.
    var removedAt: Date?
    var createdAt: Date
    var updatedAt: Date
    var writeID: UUID
    var pushedWriteID: UUID?

    var trip: TripEntity?

    /// True for synthetic identities minted without a real address (Splitwise
    /// imports, email-less Apple accounts). Never show these to users.
    var hasPlaceholderEmail: Bool { email.hasSuffix("@users.tab") }

    init(
        id: UUID = UUID(),
        userID: UUID? = nil,
        email: String,
        displayName: String,
        invitedByID: UUID? = nil,
        trip: TripEntity? = nil,
        joinedAt: Date? = nil,
        removedAt: Date? = nil,
        createdAt: Date = .now,
        updatedAt: Date = .now,
        writeID: UUID = UUID(),
        pushedWriteID: UUID? = nil
    ) {
        self.id = id
        self.userID = userID
        self.email = email
        self.displayName = displayName
        self.invitedByID = invitedByID
        self.trip = trip
        self.joinedAt = joinedAt
        self.removedAt = removedAt
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.writeID = writeID
        self.pushedWriteID = pushedWriteID
    }
}

// MARK: - Category

@Model
final class CategoryEntity {
    #Unique<CategoryEntity>([\.id])

    var id: UUID
    var tripID: UUID?         // nil for built-in defaults
    var name: String
    var icon: String
    var isDefault: Bool
    var updatedAt: Date
    var deletedAt: Date?
    var writeID: UUID
    var pushedWriteID: UUID?

    init(
        id: UUID = UUID(),
        tripID: UUID? = nil,
        name: String,
        icon: String,
        isDefault: Bool,
        updatedAt: Date = .now,
        deletedAt: Date? = nil,
        writeID: UUID = UUID(),
        pushedWriteID: UUID? = nil
    ) {
        self.id = id
        self.tripID = tripID
        self.name = name
        self.icon = icon
        self.isDefault = isDefault
        self.updatedAt = updatedAt
        self.deletedAt = deletedAt
        self.writeID = writeID
        self.pushedWriteID = pushedWriteID
    }
}

// MARK: - Expense

@Model
final class ExpenseEntity {
    #Unique<ExpenseEntity>([\.id])

    var id: UUID
    var amount: Decimal
    var currency: String
    var categoryID: UUID?
    var descriptionText: String
    var expenseDate: Date
    var receiptStoragePath: String?
    var paymentMethodRaw: String
    var createdByID: UUID
    var lastEditedByID: UUID?
    var createdAt: Date
    var updatedAt: Date
    var deletedAt: Date?
    var writeID: UUID
    var pushedWriteID: UUID?

    var trip: TripEntity?

    @Relationship(deleteRule: .cascade, inverse: \PaymentEntity.expense)
    var payments: [PaymentEntity] = []

    @Relationship(deleteRule: .cascade, inverse: \ExpenseSplitEntity.expense)
    var splits: [ExpenseSplitEntity] = []

    init(
        id: UUID = UUID(),
        amount: Decimal,
        currency: String,
        categoryID: UUID? = nil,
        descriptionText: String,
        expenseDate: Date,
        receiptStoragePath: String? = nil,
        paymentMethodRaw: String = "card",
        createdByID: UUID,
        lastEditedByID: UUID? = nil,
        trip: TripEntity? = nil,
        createdAt: Date = .now,
        updatedAt: Date = .now,
        deletedAt: Date? = nil,
        writeID: UUID = UUID(),
        pushedWriteID: UUID? = nil
    ) {
        self.id = id
        self.amount = amount
        self.currency = currency
        self.categoryID = categoryID
        self.descriptionText = descriptionText
        self.expenseDate = expenseDate
        self.receiptStoragePath = receiptStoragePath
        self.paymentMethodRaw = paymentMethodRaw
        self.createdByID = createdByID
        self.lastEditedByID = lastEditedByID
        self.trip = trip
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.deletedAt = deletedAt
        self.writeID = writeID
        self.pushedWriteID = pushedWriteID
    }
}

// MARK: - Payment

@Model
final class PaymentEntity {
    #Unique<PaymentEntity>([\.id])

    var id: UUID
    var tripPersonID: UUID
    var amountPaid: Decimal
    var paymentModeRaw: String
    var updatedAt: Date
    var writeID: UUID
    var pushedWriteID: UUID?

    var expense: ExpenseEntity?

    init(
        tripPersonID: UUID,
        amountPaid: Decimal,
        paymentModeRaw: String,
        expense: ExpenseEntity? = nil,
        updatedAt: Date = .now,
        writeID: UUID = UUID(),
        pushedWriteID: UUID? = nil
    ) {
        self.id = UUID()
        self.tripPersonID = tripPersonID
        self.amountPaid = amountPaid
        self.paymentModeRaw = paymentModeRaw
        self.expense = expense
        self.updatedAt = updatedAt
        self.writeID = writeID
        self.pushedWriteID = pushedWriteID
    }
}

// MARK: - ExpenseSplit

@Model
final class ExpenseSplitEntity {
    #Unique<ExpenseSplitEntity>([\.id])

    var id: UUID
    var tripPersonID: UUID
    var amountOwed: Decimal
    var splitTypeRaw: String
    /// Share weight when splitTypeRaw == "shares"; nil otherwise.
    var shareUnits: Decimal?
    /// Percentage of the total (0–100) when splitTypeRaw == "percentage"; nil otherwise.
    var percentage: Decimal?
    var updatedAt: Date
    var writeID: UUID
    var pushedWriteID: UUID?

    var expense: ExpenseEntity?

    init(
        tripPersonID: UUID,
        amountOwed: Decimal,
        splitTypeRaw: String,
        shareUnits: Decimal? = nil,
        percentage: Decimal? = nil,
        expense: ExpenseEntity? = nil,
        updatedAt: Date = .now,
        writeID: UUID = UUID(),
        pushedWriteID: UUID? = nil
    ) {
        self.id = UUID()
        self.tripPersonID = tripPersonID
        self.amountOwed = amountOwed
        self.splitTypeRaw = splitTypeRaw
        self.shareUnits = shareUnits
        self.percentage = percentage
        self.expense = expense
        self.updatedAt = updatedAt
        self.writeID = writeID
        self.pushedWriteID = pushedWriteID
    }
}

// MARK: - Activity (notification feed)

/// One row per trip event, mirrored from the server-side `activity_log`.
/// Append-only and immutable: the client only ever inserts new rows on pull.
@Model
final class ActivityEntity {
    #Unique<ActivityEntity>([\.id])

    var id: UUID
    var tripID: UUID
    var actorID: UUID
    var action: String
    var entityType: String
    var entityID: UUID
    var timestamp: Date
    /// Device-local read marker. Never synced; pulls only insert new activity rows.
    var readAt: Date?
    /// Device-local dismissal (swipe-to-delete). The row is kept, not deleted,
    /// so the next pull can't re-insert it; the feed just stops showing it.
    var dismissedAt: Date?
    /// JSON-encoded `[String: String]` snapshot for offline rendering + push text.
    var snapshotData: Data?

    init(
        id: UUID,
        tripID: UUID,
        actorID: UUID,
        action: String,
        entityType: String,
        entityID: UUID,
        timestamp: Date,
        readAt: Date? = nil,
        dismissedAt: Date? = nil,
        snapshotData: Data? = nil
    ) {
        self.id = id
        self.tripID = tripID
        self.actorID = actorID
        self.action = action
        self.entityType = entityType
        self.entityID = entityID
        self.timestamp = timestamp
        self.readAt = readAt
        self.dismissedAt = dismissedAt
        self.snapshotData = snapshotData
    }

    var snapshot: [String: String] {
        guard let snapshotData,
              let dict = try? JSONDecoder().decode([String: String].self, from: snapshotData)
        else { return [:] }
        return dict
    }
}

// MARK: - Trip mute preference

/// Per-trip mute for the current user, mirroring `trip_mute_prefs`.
/// `isMuted == false` is a local tombstone awaiting an unmute push.
@Model
final class TripMuteEntity {
    #Unique<TripMuteEntity>([\.tripID])

    var tripID: UUID
    var isMuted: Bool
    var mutedAt: Date
    var updatedAt: Date
    var writeID: UUID
    var pushedWriteID: UUID?

    init(
        tripID: UUID,
        isMuted: Bool = true,
        mutedAt: Date = .now,
        updatedAt: Date = .now,
        writeID: UUID = UUID(),
        pushedWriteID: UUID? = nil
    ) {
        self.tripID = tripID
        self.isMuted = isMuted
        self.mutedAt = mutedAt
        self.updatedAt = updatedAt
        self.writeID = writeID
        self.pushedWriteID = pushedWriteID
    }
}

// MARK: - Settlement

@Model
final class SettlementEntity {
    #Unique<SettlementEntity>([\.id])

    var id: UUID
    var fromPersonID: UUID
    var toPersonID: UUID
    var amount: Decimal
    var currency: String
    var note: String?
    var settledAt: Date
    var createdByID: UUID
    var createdAt: Date
    var updatedAt: Date
    var deletedAt: Date?
    var writeID: UUID
    var pushedWriteID: UUID?

    var trip: TripEntity?

    init(
        id: UUID = UUID(),
        fromPersonID: UUID,
        toPersonID: UUID,
        amount: Decimal,
        currency: String,
        note: String? = nil,
        settledAt: Date,
        createdByID: UUID,
        trip: TripEntity? = nil,
        createdAt: Date = .now,
        updatedAt: Date = .now,
        deletedAt: Date? = nil,
        writeID: UUID = UUID(),
        pushedWriteID: UUID? = nil
    ) {
        self.id = id
        self.fromPersonID = fromPersonID
        self.toPersonID = toPersonID
        self.amount = amount
        self.currency = currency
        self.note = note
        self.settledAt = settledAt
        self.createdByID = createdByID
        self.trip = trip
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.deletedAt = deletedAt
        self.writeID = writeID
        self.pushedWriteID = pushedWriteID
    }
}

File: Apps/TabAndroid/core/data/src/main/kotlin/com/rithwikshetty/tab/data/local/Entities.kt
package com.rithwikshetty.tab.data.local

import androidx.room.ColumnInfo
import androidx.room.Embedded
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey

public data class SyncStamp(
    @ColumnInfo(name = "updated_at")
    public val updatedAt: String,
    @ColumnInfo(name = "deleted_at")
    public val deletedAt: String?,
    @ColumnInfo(name = "write_id")
    public val writeId: String,
    @ColumnInfo(name = "is_dirty")
    public val isDirty: Boolean,
)

@Entity(tableName = "profiles")
public data class ProfileEntity(
    @PrimaryKey
    public val id: String,
    @ColumnInfo(name = "display_name")
    public val displayName: String,
    @ColumnInfo(name = "avatar_url")
    public val avatarUrl: String?,
    @ColumnInfo(name = "activity_last_seen_at")
    public val activityLastSeenAt: String?,
    @ColumnInfo(name = "created_at")
    public val createdAt: String,
    @Embedded
    public val sync: SyncStamp,
)

@Entity(
    tableName = "trips",
    foreignKeys = [
        ForeignKey(
            entity = ProfileEntity::class,
            parentColumns = ["id"],
            childColumns = ["created_by"],
            onDelete = ForeignKey.RESTRICT,
        ),
    ],
    indices = [
        Index(value = ["created_by"]),
        Index(value = ["last_activity_at"]),
        Index(value = ["member_signature"], unique = true),
    ],
)
public data class TripEntity(
    @PrimaryKey
    public val id: String,
    public val name: String,
    public val kind: String,
    @ColumnInfo(name = "member_signature")
    public val memberSignature: String?,
    @ColumnInfo(name = "created_by")
    public val createdBy: String,
    @ColumnInfo(name = "last_activity_at")
    public val lastActivityAt: String,
    @ColumnInfo(name = "created_at")
    public val createdAt: String,
    @Embedded
    public val sync: SyncStamp,
)

@Entity(
    tableName = "trip_people",
    primaryKeys = ["id"],
    foreignKeys = [
        ForeignKey(
            entity = TripEntity::class,
            parentColumns = ["id"],
            childColumns = ["trip_id"],
            onDelete = ForeignKey.CASCADE,
        ),
    ],
    indices = [
        Index(value = ["trip_id", "email"], unique = true),
        Index(value = ["trip_id", "user_id"], unique = true),
        Index(value = ["user_id"]),
    ],
)
public data class TripPersonEntity(
    public val id: String,
    @ColumnInfo(name = "trip_id")
    public val tripId: String,
    @ColumnInfo(name = "user_id")
    public val userId: String?,
    public val email: String,
    @ColumnInfo(name = "display_name")
    public val displayName: String,
    @ColumnInfo(name = "invited_by")
    public val invitedBy: String?,
    @ColumnInfo(name = "joined_at")
    public val joinedAt: String?,
    @ColumnInfo(name = "removed_at")
    public val removedAt: String?,
    @ColumnInfo(name = "created_at")
    public val createdAt: String,
    @Embedded
    public val sync: SyncStamp,
)

@Entity(
    tableName = "categories",
    foreignKeys = [
        ForeignKey(
            entity = TripEntity::class,
            parentColumns = ["id"],
            childColumns = ["trip_id"],
            onDelete = ForeignKey.CASCADE,
        ),
    ],
    indices = [Index(value = ["trip_id"])],
)
public data class CategoryEntity(
    @PrimaryKey
    public val id: String,
    @ColumnInfo(name = "trip_id")
    public val tripId: String?,
    public val name: String,
    public val icon: String,
    @ColumnInfo(name = "is_default")
    public val isDefault: Boolean,
    @ColumnInfo(name = "created_at")
    public val createdAt: String,
    @Embedded
    public val sync: SyncStamp,
)

@Entity(
    tableName = "expenses",
    foreignKeys = [
        ForeignKey(
            entity = TripEntity::class,
            parentColumns = ["id"],
            childColumns = ["trip_id"],
            onDelete = ForeignKey.RESTRICT,
        ),
        ForeignKey(
            entity = CategoryEntity::class,
            parentColumns = ["id"],
            childColumns = ["category_id"],
            onDelete = ForeignKey.SET_NULL,
        ),
        ForeignKey(
            entity = ProfileEntity::class,
            parentColumns = ["id"],
            childColumns = ["created_by"],
            onDelete = ForeignKey.RESTRICT,
        ),
    ],
    indices = [
        Index(value = ["trip_id", "expense_date"]),
        Index(value = ["category_id"]),
        Index(value = ["created_by"]),
    ],
)
public data class ExpenseEntity(
    @PrimaryKey
    public val id: String,
    @ColumnInfo(name = "trip_id")
    public val tripId: String,
    @ColumnInfo(name = "amount_text")
    public val amountText: String,
    public val currency: String,
    @ColumnInfo(name = "category_id")
    public val categoryId: String?,
    public val description: String,
    @ColumnInfo(name = "expense_date")
    public val expenseDate: String,
    @ColumnInfo(name = "receipt_storage_path")
    public val receiptStoragePath: String?,
    @ColumnInfo(name = "payment_method")
    public val paymentMethod: String,
    @ColumnInfo(name = "created_by")
    public val createdBy: String,
    @ColumnInfo(name = "last_edited_by")
    public val lastEditedBy: String?,
    @ColumnInfo(name = "created_at")
    public val createdAt: String,
    @Embedded
    public val sync: SyncStamp,
)

@Entity(
    tableName = "expense_payments",
    primaryKeys = ["expense_id", "trip_person_id"],
    foreignKeys = [
        ForeignKey(
            entity = ExpenseEntity::class,
            parentColumns = ["id"],
            childColumns = ["expense_id"],
            onDelete = ForeignKey.CASCADE,
        ),
        ForeignKey(
            entity = TripPersonEntity::class,
            parentColumns = ["id"],
            childColumns = ["trip_person_id"],
            onDelete = ForeignKey.RESTRICT,
        ),
    ],
    indices = [Index(value = ["trip_person_id"])],
)
public data class ExpensePaymentEntity(
    @ColumnInfo(name = "expense_id")
    public val expenseId: String,
    @ColumnInfo(name = "trip_person_id")
    public val tripPersonId: String,
    @ColumnInfo(name = "amount_paid_text")
    public val amountPaidText: String,
    @ColumnInfo(name = "payment_mode")
    public val paymentMode: String,
    @ColumnInfo(name = "created_at")
    public val createdAt: String,
    @Embedded
    public val sync: SyncStamp,
)

@Entity(
    tableName = "expense_splits",
    primaryKeys = ["expense_id", "trip_person_id"],
    foreignKeys = [
        ForeignKey(
            entity = ExpenseEntity::class,
            parentColumns = ["id"],
            childColumns = ["expense_id"],
            onDelete = ForeignKey.CASCADE,
        ),
        ForeignKey(
            entity = TripPersonEntity::class,
            parentColumns = ["id"],
            childColumns = ["trip_person_id"],
            onDelete = ForeignKey.RESTRICT,
        ),
    ],
    indices = [Index(value = ["trip_person_id"])],
)
public data class ExpenseSplitEntity(
    @ColumnInfo(name = "expense_id")
    public val expenseId: String,
    @ColumnInfo(name = "trip_person_id")
    public val tripPersonId: String,
    @ColumnInfo(name = "amount_owed_text")
    public val amountOwedText: String,
    @ColumnInfo(name = "split_type")
    public val splitType: String,
    @ColumnInfo(name = "share_units_text")
    public val shareUnitsText: String?,
    @ColumnInfo(name = "percentage_text")
    public val percentageText: String?,
    @ColumnInfo(name = "created_at")
    public val createdAt: String,
    @Embedded
    public val sync: SyncStamp,
)

@Entity(
    tableName = "settlements",
    foreignKeys = [
        ForeignKey(
            entity = TripEntity::class,
            parentColumns = ["id"],
            childColumns = ["trip_id"],
            onDelete = ForeignKey.RESTRICT,
        ),
        ForeignKey(
            entity = TripPersonEntity::class,
            parentColumns = ["id"],
            childColumns = ["from_person_id"],
            onDelete = ForeignKey.RESTRICT,
        ),
        ForeignKey(
            entity = TripPersonEntity::class,
            parentColumns = ["id"],
            childColumns = ["to_person_id"],
            onDelete = ForeignKey.RESTRICT,
        ),
    ],
    indices = [
        Index(value = ["trip_id", "settled_at"]),
        Index(value = ["from_person_id"]),
        Index(value = ["to_person_id"]),
    ],
)
public data class SettlementEntity(
    @PrimaryKey
    public val id: String,
    @ColumnInfo(name = "trip_id")
    public val tripId: String,
    @ColumnInfo(name = "from_person_id")
    public val fromPersonId: String,
    @ColumnInfo(name = "to_person_id")
    public val toPersonId: String,
    @ColumnInfo(name = "amount_text")
    public val amountText: String,
    public val currency: String,
    public val note: String?,
    @ColumnInfo(name = "settled_at")
    public val settledAt: String,
    @ColumnInfo(name = "created_by")
    public val createdBy: String,
    @ColumnInfo(name = "created_at")
    public val createdAt: String,
    @Embedded
    public val sync: SyncStamp,
)

@Entity(
    tableName = "activity_log",
    indices = [
        Index(value = ["trip_id", "timestamp"]),
        Index(value = ["entity_type", "entity_id"]),
    ],
)
public data class ActivityEntity(
    @PrimaryKey
    public val id: String,
    @ColumnInfo(name = "trip_id")
    public val tripId: String,
    @ColumnInfo(name = "actor_id")
    public val actorId: String,
    public val action: String,
    @ColumnInfo(name = "entity_type")
    public val entityType: String,
    @ColumnInfo(name = "entity_id")
    public val entityId: String,
    public val timestamp: String,
    @ColumnInfo(name = "snapshot_json")
    public val snapshotJson: String?,
)

@Entity(
    tableName = "trip_mute_preferences",
    primaryKeys = ["trip_id", "user_id"],
    foreignKeys = [
        ForeignKey(
            entity = TripEntity::class,
            parentColumns = ["id"],
            childColumns = ["trip_id"],
            onDelete = ForeignKey.CASCADE,
        ),
    ],
    indices = [Index(value = ["user_id"])],
)
public data class TripMutePreferenceEntity(
    @ColumnInfo(name = "trip_id")
    public val tripId: String,
    @ColumnInfo(name = "user_id")
    public val userId: String,
    @ColumnInfo(name = "muted_at")
    public val mutedAt: String,
    @Embedded
    public val sync: SyncStamp,
)

@Entity(
    tableName = "receipt_drafts",
    foreignKeys = [
        ForeignKey(
            entity = ExpenseEntity::class,
            parentColumns = ["id"],
            childColumns = ["expense_id"],
            onDelete = ForeignKey.CASCADE,
        ),
    ],
    indices = [Index(value = ["upload_state"])],
)
public data class ReceiptDraftEntity(
    @PrimaryKey
    @ColumnInfo(name = "expense_id")
    public val expenseId: String,
    @ColumnInfo(name = "local_uri")
    public val localUri: String,
    @ColumnInfo(name = "remote_path")
    public val remotePath: String?,
    @ColumnInfo(name = "upload_state")
    public val uploadState: String,
    @ColumnInfo(name = "updated_at")
    public val updatedAt: String,
)

@Entity(
    tableName = "sync_outbox",
    indices = [
        Index(value = ["entity_type", "entity_id", "write_id"], unique = true),
        Index(value = ["next_attempt_at", "sequence"]),
    ],
)
public data class OutboxEntity(
    @PrimaryKey(autoGenerate = true)
    public val sequence: Long = 0,
    @ColumnInfo(name = "entity_type")
    public val entityType: String,
    @ColumnInfo(name = "entity_id")
    public val entityId: String,
    public val operation: String,
    @ColumnInfo(name = "write_id")
    public val writeId: String,
    @ColumnInfo(name = "created_at")
    public val createdAt: String,
    @ColumnInfo(name = "attempt_count")
    public val attemptCount: Int = 0,
    @ColumnInfo(name = "last_error")
    public val lastError: String? = null,
    @ColumnInfo(name = "next_attempt_at")
    public val nextAttemptAt: String? = null,
)
